server/extensions/handler.go

```diff
@@ -0,0 +1,182 @@
+			return nil, err
+		}
+		if err := yaml.UnmarshalStrict(i.Data["paths"], &paths); err != nil {
+			return nil, err
+		}
+
+		basePath := fmt.Sprintf("/api/extensions/%s", name)
+
+		// prevent a UI extension accidentally DoS the Argo CD Server
+		// one rate-limiter per extension, means one bad extension does not impact others
+		limiter := rate.NewLimiter(1, 3)
+
+		for subPath, methods := range paths {
+			path := basePath + subPath
+			log.
+				WithField("name", name).
+				WithField("path", path).
+				Info("Loading API extension")
+
+			r.HandleFunc(path, func(w http.ResponseWriter, r *http.Request) {
+				ctx := r.Context()
+
+				// rate-limit the request
+				if !limiter.Allow() {
+					http.Error(w, http.StatusText(http.StatusTooManyRequests), http.StatusTooManyRequests)
+					return
+				}
+
+				// get the claims from the context
+				claims, ok := ctx.Value("claims").(*jwt.MapClaims)
+				if !ok {
+					http.Error(w, http.StatusText(http.StatusUnauthorized), http.StatusUnauthorized)
+					return
+				}
+
+				// is the method allowed?
+				m, ok := methods[r.Method]
+				if !ok {
+					http.Error(w, http.StatusText(http.StatusMethodNotAllowed), http.StatusMethodNotAllowed)
+					return
+				}
+
+				// get the object from the request
+				vars := mux.Vars(r)
+				object, ok := vars[m.Object]
+				if !ok {
+					http.Error(w, http.StatusText(http.StatusBadRequest)+": missing object", http.StatusBadRequest)
+					return
+				}
+
+				// create a conventional name for the object
+				object, err := rbacObject(ctx, m.Resource, object)
+				if err != nil {
+					http.Error(w, http.StatusText(http.StatusServiceUnavailable)+": object lookup failed", http.StatusServiceUnavailable)
+					return
+				}
+
+				// enforce access controls
+				if err := enforcer.EnforceErr(claims, m.Resource, m.Action, object); err != nil {
+					http.Error(w, http.StatusText(http.StatusForbidden), http.StatusForbidden)
+					return
+				}
+
+				// proxy the request
+				req, err := http.NewRequestWithContext(
+					// use the HTTP request's context, if that is cancelled, this will be canceled to
+					ctx,
+					r.Method,
+					url+strings.TrimPrefix(r.URL.Path, basePath),
+					r.Body,
+				)
+				if err != nil {
+					http.Error(w, err.Error(), http.StatusInternalServerError)
+					return
+				}
+				req.Header = headers.Clone()
+
+				// we need to provide information to the downstream about who is making the request
+				req.Header.Add("claims-sub", (*claims)["sub"].(string))
+
+				log.WithField("sub", req.Header.Get("claims-sub")).
+					WithField("method", req.Method).
+					WithField("url", req.URL).
+					Info("Executing API extension request")
+
+				resp, err := httpClient.Do(req)
+				if err != nil {
+					http.Error(w, err.Error(), http.StatusBadGateway)
+					return
+				}
+				defer resp.Body.Close()
+
+				w.WriteHeader(resp.StatusCode)
+				for k, v := range resp.Header {
+					w.Header()[k] = v
+				}
```

Comment:
Should we be concerned that the response headers might include a secret which should not be made available to the user?